### PR TITLE
Add Preferences library API and tutorial documents

### PR DESCRIPTION
**Pull request by Xylopyrographer**
## Summary
Adds API and tutorial documents for the  Preferences library.

## Impact
None

### CODE REVIEW

**Comment by Xylopyrographer** [comment]
@pedrominatel Since I never seem to get PR's right the first time, could you take a look at  this and let me know if I need to do anything else to move this along. Thanks.

File: docs/source/api/preferences.rst [hunk 1]
```diff
@@ -0,0 +1,702 @@
...
+The Preferences library is unique to arduino-esp32. It should be considered as the replacement for the Arduino EEPROM library.
+
+It uses a portion of the on-board non-volatile memory (NVS) of the ESP32 to store data. This data is retained across restarts and loss of power events to the system. 
+
+Preferences works best for storing many small values, rather than a few large values. If large amounts of data are to be stored, consider using a file system library such as LitteFS.
+
+The Preferences library is usable by all ESP32 variants.
+
+
+Header file
```
Comment (pedrominatel) [inline docs/source/api/preferences.rst:19]: For titles and heading, we use capital letters for the first letter of each word. Exceptions are: is, of, or, for, and...

File: docs/source/tutorials/preferences.rst [hunk 2]
```diff
@@ -0,0 +1,699 @@
...
+
+   Preferences mySketchPrefs;
+   
+   mySketchPrefs.begin("myPrefs", false);
+   size_t whatsLeft = freeEntries();    // this method works regardless of the mode in which the namespace is opened.
+   Serial.printf("There are: %u entries available in the namespace table.\n, whatsLeft);
+   
+..
+
+will print to the serial monitor this information.
```
Comment (pedrominatel) [inline docs/source/tutorials/preferences.rst:386]: This part looks strange. Maybe is broken somehow.

File: docs/source/api/preferences.rst [hunk 3]
```diff
@@ -0,0 +1,702 @@
+.. 2022-03-17 by Xylopyrographer. Initial PR release.
```
Comment (pedrominatel) [inline docs/source/api/preferences.rst:1]: We do not add the authors' name on the docs. Please consider removing it.

Comment (pedrominatel) [reply docs/source/api/preferences.rst:1]: You will be already added to the contributor's list and in the commit.

File: docs/source/tutorials/preferences.rst [hunk 4]
```diff
@@ -0,0 +1,699 @@
...
+
+Fortunately this is not as difficult as it may sound as the ``getBytesLength`` method and the ``sizeof`` operator help with keeping track of it all.
+
+This is best explained with an example. Here we use the ``Bytes`` methods to store and retrieve an array, while ensuring the data type is preserved.
+
+.. code-block:: arduino
+
+   /*
+    *  An example sketch using the Preferences "Bytes" methods
+    *   to store and retreive an arbitrary number of bytes in
```
Comment (pedrominatel) [inline docs/source/tutorials/preferences.rst:447]: retreive -> retrieve

**Review by pedrominatel — APPROVED** [review]
LGTM

**Comment by Xylopyrographer** [comment]
@pedrominatel 
One thing the might need fixing. In the tutorial preferences.rst is a link to the API preferences.rst document. I don't know what the proper syntax is for creating this link so I totally guessed at one.

File: docs/source/tutorials/preferences.rst [hunk 5]
```diff
@@ -0,0 +1,697 @@
...
+.. 
+
+.. warning:: 
+   **You should download a new sketch to your board immediately after running the above or else it will reformat the NVS partition every time it is powered up or restarted!**
+
+
+Resources
+---------
+
+* `Preferences API document`_
```
Comment (pedrominatel) [inline docs/source/tutorials/preferences.rst:675]: Change to: 
```
`Preferences API Reference <../api/preferences.html>`_
```

File: docs/source/tutorials/preferences.rst [hunk 6]
```diff
@@ -0,0 +1,697 @@
...
+* `Non-volatile storage library`_ (espressif-IDF API Reference)
+* `Official ESP-IDF documentation`_ (espressif-IDF Reference)
+
+.. _Preferences API document: https://../api/preferences.html
```
Comment (pedrominatel) [inline docs/source/tutorials/preferences.rst:679]: Remove this line.

**Review by pedrominatel — COMMENTED** [review]
PTAL

**Review by VojtechBartoska — APPROVED** [review]
I went through it quickly and it looks good to me. Thanks for your extensive contribution @Xylopyrographer!

File: docs/source/tutorials/preferences.rst (hunk 5 above)
Comment (Xylopyrographer) [reply docs/source/tutorials/preferences.rst]: 👍 Done.

**Pull request merged**